src/auth.rs

```diff
@@ -0,0 +1,115 @@
+use cached::proc_macro::cached;
+use serde::{Deserialize, Serialize};
+
+use crate::{ApiEnvironment, ApiError, Mpesa, MpesaError, MpesaResult};
+
+const AUTHENTICATION_URL: &str = "/oauth/v1/generate?grant_type=client_credentials";
+
+/// Response returned from the authentication function
+#[derive(Debug, Serialize, Deserialize)]
+pub struct AuthenticationResponse {
+    /// Access token which is used as the Bearer-Auth-Token
+    pub access_token: String,
+    /// Expiry time in seconds
+    pub expiry_in: u64,
+}
+
+impl std::fmt::Display for AuthenticationResponse {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "token :{} expires in: {}",
+            self.access_token, self.expiry_in
+        )
+    }
+}
+
+#[cached(
+    size = 1,
+    time = 3600,
+    key = "String",
+    result = true,
+    convert = r#"{ format!("{}", client.client_key()) }"#
+)]
+pub(crate) async fn auth(client: &Mpesa<impl ApiEnvironment>) -> MpesaResult<String> {
+    let url = format!("{}{}", client.environment.base_url(), AUTHENTICATION_URL);
+
+    let response = client
+        .http_client
+        .get(&url)
+        .basic_auth(client.client_key(), Some(&client.client_secret()))
+        .send()
+        .await?;
+
+    if response.status().is_success() {
+        let value = response.json::<AuthenticationResponse>().await?;
+        let access_token = value.access_token;
+
+        return Ok(access_token);
+    }
```

Comment:
Instead of having to check for `response.status().is_success()` we can convert http responses that are not `20*` to `reqwest::Error` by calling  the `error_for_status` after sending the request

```suggestion
    let response = client
        .http_client
        .get(&url)
        .basic_auth(client.client_key(), Some(&client.client_secret()))
        .send()
        .await?
        .error_for_status()?
        .json::<AuthenticationResponse>()
        .await?;
    Ok(response.access_token)
```

Reply:
I am seeing with this approach we could lose the ability to return a `MpesaError::AuthenticationError` on a non `20*` response. Perhaps mapping the error from `error_for_status` could fix this, otherwise, the verbose code could suffice.